### PR TITLE
Add an option to allow passing ignored_parameters to FSDP

### PR DESCRIPTION
**Pull request by zhaojuanmao**
This is mainly to resolve composability issue btw TorchRec and FSDP. 

For some types of sharded modules in TorchRec, their named_parameters() and named_modules() are overridden by TorchRec and do not follow the PyTorch named_parameters() and named_modules() convention. 

For such cases, ignored_modules do not work well as the parameters extracted from ignored_modules() may not match sparse sharded parameters in TorchRec.dmp_wrapped_module.parameters(). 

One solution is to allow users to pass ignored_parameters directly to FSDP.  e.g. they can call sharded_parameter_names() API in TorchRec to get the right ignored_parameters and then pass them to FSDP. The parameters returned from sharded_parameter_names() will match sparse sharded parameters in TorchRec.dmp_wrapped_module..parameters(), as both of they are offerred/overriden by TorchRec.

In this PR, we do not allow users to pass both ignored_modules and ignored_parameters for simplicity (in theory, it should work, but no need to add complexity for tests and possibly edge cases? especially we do not have a real use case to use both yet).

### CODE REVIEW

**Comment by zhaojuanmao** [comment]
actually we want to support passing ignore_parameter_names..., will update PR

**Comment by awgu** [comment]
I want to discuss whether we want a separate constructor argument for `ignored_parameter_names` or whether we should unify with `ignored_modules` into one argument.

There are two possible approaches to unify: `ignored_states: Optional[List[Union[nn.Module, str]]]` and `ignored_states: Optional[Union[List[nn.Module], List[str]]]`, where in both cases FSDP's internal logic will process the `ignored_states` as needed.

1. To clarify, do you expect users to:
- pass `ignored_modules` without `ignored_parameter_names`? (answer should be yes since this is what exists today)
- pass `ignored_parameter_names` without `ignored_modules`? (answer seems like yes for your mentioned TorchRec use case)
- pass `ignored_modules` and `ignored_parameter_names`? (your PR description suggests that this is probably not needed)

If passing `ignored_modules` and `ignored_parameter_names` together is not expected, maybe we should not expose that option to users publicly in the first place. In that case, we prefer a unified argument like `ignored_states: Optional[Union[List[nn.Module], [str]]]`, where the user may _either_ pass a list of modules _or_ a list of parameter names. The type annotation plus our public documentation can make this explicit, and FSDP's internal logic can validate the input.

2. More broadly, do you expect `ignored_parameter_names` to be a long-term argument to `fully_shard`? We previously discussed removing `ignored_modules` from `fully_shard` altogether since it deviates from the spirit of the composable APIs. A `keep_local` annotation may be preferred to handle ignored modules. 

We should be careful about the `fully_shard` constructor especially as we are starting to integrate internally. Since next half we want to simplify even `FullyShardedDataParallel`'s constructor, then we should begin thinking about the ideal constructor.

One minor challenge is that if the `FullyShardedDataParallel` constructor arguments differs from those of `fully_shard`, then switching between the two may be less convenient. However, I think strictly restricting the two to have the same constructor arguments is counter to the purpose of  `fully_shard`. (Just as a reminder, I already use `policy` instead of `auto_wrap_policy` for `fully_shard` since `fully_shard` does not wrap and moreover the argument could be used to toggle other behavior such as tensor-shape preserving sharding in the future.)

**Comment by zhaojuanmao** [comment]
1. I think we should not encourage users to use both ignored_modules and ignored_parameter_names. ignored_parameter_names is actually more general than ignored_modules. That being said, the cases who use ignored_modules can migrate to use ignored_parameter_names as well.  we keep both for back compatibility for now, and we can migrate internal use  cases to use ignored_parameter_names once the PR is landed. after that we can kill ignored_modules. How do you think?
2. agree that we should keep fully_shard() simple and no need to be exactly same as FullyShardedDataParallel, as long as keep_local() can handle the cases that use ignored_parameter_names in the future, we should be good. 

by the way, I'm not using ghstack, so this PR has the changes in the PR https://github.com/pytorch/pytorch/pull/91530, it is better to review and land https://github.com/pytorch/pytorch/pull/91530 first. 

Thanks!

File: torch/distributed/_composable/fully_shard.py
Comment (rohan-varma) [inline torch/distributed/_composable/fully_shard.py:48]: nit: add docstring and example

File: torch/distributed/fsdp/_init_utils.py
Comment (rohan-varma) [inline torch/distributed/fsdp/_init_utils.py:611]: is it possible to pass an FSDP flattened ignored parameter here and have it indeed be filtered out? Can we add a test that would simulate this?

File: torch/distributed/fsdp/_init_utils.py
Comment (rohan-varma) [inline torch/distributed/fsdp/_init_utils.py:613]: in theory exactly only one of these (or neither I guess) should be populated right, since we only allow passing one or the other for now?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:335]: nit: add docstring

File: torch/distributed/fsdp/_state_dict_utils.py
Comment (rohan-varma) [inline torch/distributed/fsdp/_state_dict_utils.py:310]: curious why we're removing `_ignored_param_names`?

**Review by rohan-varma — APPROVED** [review]
Stamping to unblock, but I think we should align on what final API/contract we want to offer,  and definitely add unittests before landing. 

Re: docstring, if we don't want this to be public facing API, wonder if we can prefix ctor arg with `_`, or have a private API to set this. Adding it directly to ctor might involve some BC concerns.

**Comment by zhaojuanmao** [comment]
> actually we want to support passing ignore_parameter_names..., will update PR

Actually it is not feasible to pass ''ignore_parameter_names' for manually nest wrapped FSDP. If we require ''ignore_parameter_names' to be fqns with prefix starting from root model, the nested FSDP instance has no way to find the parameters based on these ignored parameter names. If we require ''ignore_parameter_names' to be fqns with prefix starting from the local root module wrapped by the nested FSDP instance, then every FSDP instance requires to pass different set of ''ignore_parameter_names', which does not have good usability. 

So this PR still keeps using 'ignored_parameters' instead of 'ignored_parameter_names'

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (zhaojuanmao) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:409]: remove this, as we need to get the '_ignored_params' of nested FSDP instance. This should not result in memory leak...? as they are part of and referenced by module parameters() anyway during lifetime of training

File: torch/distributed/fsdp/_init_utils.py
Comment (zhaojuanmao) [reply torch/distributed/fsdp/_init_utils.py]: yes, the unit tests have already covered this case

File: torch/distributed/fsdp/_init_utils.py
Comment (zhaojuanmao) [reply torch/distributed/fsdp/_init_utils.py]: yes, the unit tests have already covered this case

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:320]: nit: Double backtick for code type in public docstrings (after the type annotation)
```suggestion
            argument, ``ignored_modules`` could be deprecated soon. For backward compatibility,
```

Single backtick only shows as italics:
Screen Shot 2023-01-04 at 4 32 45 AM
Example with double backtick (`False`, `True`, `bool`):
Screen Shot 2023-01-04 at 4 34 00 AM

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:322]: nit: Just want to clarify that this is a requirement, not a suggestion
```suggestion
            but FSDP only allows one of them to be specified as not ``None``.
```

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:409]: Yes, this should be fine. `use_orig_params=True` path assumes that the user does not swap parameter variables after initialization, which is the same assumption here to not have a memory leak.

File: test/distributed/fsdp/test_fsdp_ignored_modules.py
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_ignored_modules.py:260]: nit: To avoid duplication
```
ignore_kwargs = (
    {"ignored_modules": ignored_modules}
    if ignore_modules
    else {"ignored_parameters": set(p for m in ignored_modules for p in m.parameters())}
)
model.layer1 = FSDP(
    model.layer1,
    # sharding_strategy shouldn't matter here.
    sharding_strategy=ShardingStrategy.SHARD_GRAD_OP,
    **ignore_kwargs,
)
model.layer3 = FSDP(
    model.layer3,
    # the ignored modules/parameters contains submodule under model.layer1, which
    # is out of the local root model.layer3.
    **ignore_kwargs,
)
```

File: torch/distributed/fsdp/_init_utils.py
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:617]: Before this `update()`, what should be in `all_ignored_params`?

IIUC, this should only contain parameters from composable APIs that are incompatible with `fully_shard` (e.g. `replicate`) because those are the only ignored modules manually added into the `ignored_modules` set, not specified by the user. This means that for the wrapper FSDP path, `all_ignored_params` should be empty.

I want to double check if that is your understanding as well.

File: test/distributed/fsdp/test_fsdp_ignored_modules.py
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_ignored_modules.py:260]: nit: We can do something something similar for the other unit tests, but it does not need to be changed for this PR. The idea is that we can push the `if ignore_modules` / `else` branch to the tightest scope possible to avoid duplication. That way, other kwargs like `use_orig_params`, etc. do not need to be repeated. Otherwise, if we ever add a new kwargs, we will need to add it to two FSDP constructor calls per unit test.

**Review by awgu — APPROVED** [review]
This PR looks good to me!

**Comment by colin2328** [comment]
@zhaojuanmao  is this no longer needed with https://github.com/pytorch/pytorch/pull/91663 ?

**Comment by awgu** [comment]
> @zhaojuanmao is this no longer needed with #91663 ?

We discussed offline, and @zhaojuanmao accidentally landed the changes in this PR with the one you linked. She plans to follow-up on the feedback here soon.

**Comment by zhaojuanmao** [comment]
sorry for confusion, since the main codes were landed into https://github.com/pytorch/pytorch/pull/91663, I'm closing this PR and will create a new one with follow up comments (to avoid merge conflicts, etc)

File: torch/distributed/fsdp/_init_utils.py
Comment (zhaojuanmao) [reply torch/distributed/fsdp/_init_utils.py:617]: That is also my understanding based on current code reading

**Pull request closed without merging**
